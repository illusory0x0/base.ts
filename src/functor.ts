export type Instances<Instance extends Record<string, Function>> = Record<
  string,
  Instance
>;

export interface Functor {
  instances: Instances<{ map: Function }>;
}

export let Functor: Functor = {
  instances: {},
  map: (xs: object, f: any): any => {
    const xs_Functor = Functor.instances[xs.constructor.name];
    if (xs_Functor !== undefined) {
      return xs_Functor.map(xs, f);
    } else {
      throw new Error("No instance");
    }
  },
};
