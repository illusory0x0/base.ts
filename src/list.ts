import { Functor } from "./functor";
type Constructor<name extends string> = {
  readonly constructor: {
    readonly name: name;
  };
};

export type Nil = Constructor<"List"> & { readonly tag: "Nil" };

export type Cons<a> = Constructor<"List"> & {
  readonly tag: "Cons";
  readonly head: a;
  readonly tail: List<a>;
};

export type List<a> = Nil | Cons<a>;

export let Nil: { tag: "Nil" } & Constructor<"List"> = {
  tag: "Nil",
  constructor: {
    name: "List",
  },
};

export let Cons = <a>(x: a, xs: List<a>): List<a> => {
  return {
    tag: "Cons",
    constructor: {
      name: "List",
    },
    head: x,
    tail: xs,
  };
};

export let List: {
  readonly name: "List";
  readonly map: <a, b>(xs: List<a>, f: (x: a) => b) => List<b>;
  readonly from_array: <a>(xs: a[]) => List<a>;
} = {
  name: "List",
  map: (xs, f) => {
    if (xs.tag === "Nil") {
      return Nil;
    } else {
      return Cons(f(xs.head), List.map(xs.tail, f));
    }
  },
  from_array: <a>(xs: a[]) => {
    let result: List<a> = Nil;
    for (let i = 0; i !== xs.length; i++) {
      let e = xs[xs.length - 1 - i] as a;
      result = Cons(e, result);
    }
    return result;
  },
};

declare module "./functor" {
  interface Functor {
    map<a, b>(xs: List<a>, f: (x: a) => b): List<b>;
  }
}

Functor.instances[List.name] = {
  map: List.map,
};
