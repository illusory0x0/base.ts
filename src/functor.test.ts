import { test, expect } from "vitest";
import { Functor } from "./functor";

export { Array } from "./array";
export { Map } from "./map";
import { List } from "./list";

test("functor", () => {
  let xs = [1, 2, 3];
  let ys = new Map([
    ["apple", "100.23"],
    ["orange", "24.5"],
  ]);

  expect(Functor.map(xs, (x) => x.toString())).toMatchSnapshot();
  expect(Functor.map(ys, (x) => Number.parseFloat(x))).toMatchSnapshot();
  let zs = List.from_array([1, 2, 3, 4, 5]);

  expect(zs).toMatchSnapshot();
  expect(Functor.map(zs, (x) => x.toString())).toMatchSnapshot();
});
