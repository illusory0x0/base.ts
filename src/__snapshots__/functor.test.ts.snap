// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`functor 1`] = `
[
  "1",
  "2",
  "3",
]
`;

exports[`functor 2`] = `
Map {
  "apple" => 100.23,
  "orange" => 24.5,
}
`;

exports[`functor 3`] = `
{
  "constructor": {
    "name": "List",
  },
  "head": 1,
  "tag": "Cons",
  "tail": {
    "constructor": {
      "name": "List",
    },
    "head": 2,
    "tag": "Cons",
    "tail": {
      "constructor": {
        "name": "List",
      },
      "head": 3,
      "tag": "Cons",
      "tail": {
        "constructor": {
          "name": "List",
        },
        "head": 4,
        "tag": "Cons",
        "tail": {
          "constructor": {
            "name": "List",
          },
          "head": 5,
          "tag": "Cons",
          "tail": {
            "constructor": {
              "name": "List",
            },
            "tag": "Nil",
          },
        },
      },
    },
  },
}
`;

exports[`functor 4`] = `
{
  "constructor": {
    "name": "List",
  },
  "head": "1",
  "tag": "Cons",
  "tail": {
    "constructor": {
      "name": "List",
    },
    "head": "2",
    "tag": "Cons",
    "tail": {
      "constructor": {
        "name": "List",
      },
      "head": "3",
      "tag": "Cons",
      "tail": {
        "constructor": {
          "name": "List",
        },
        "head": "4",
        "tag": "Cons",
        "tail": {
          "constructor": {
            "name": "List",
          },
          "head": "5",
          "tag": "Cons",
          "tail": {
            "constructor": {
              "name": "List",
            },
            "tag": "Nil",
          },
        },
      },
    },
  },
}
`;
